### PR TITLE
Adding match-hosts, match-ports and match-paths.

### PR DESCRIPTION
**Pull request by Sun77789**
@mikewest

### CODE REVIEW

File: index.src.html [hunk 1]
```diff
@@ -3543,6 +3483,103 @@ <h5 id="effective-directive-for-a-request" algorithm>
...
+
+      1.  If B is the <a>default port</a> for Scheme-B, return "`Matches`".
+          
+      2.  Return "`Does Not Match`".
+
+  2.  If A is equal to ":*", return "`Matches`".
+
+  3.  If A is equal to B, return "`Matches`".
+
+  4.  If B is empty:
```
Comment (Sun77789) [inline index.src.html]: Should I change this case back to "If expression’s port-part is 80, and url’s port is 443"? 
I thought it might be more clear to change it to default port since this is supposed to be a generic algorithm?

File: index.src.html [hunk 2]
```diff
@@ -3371,79 +3371,19 @@ <h5 id="match-url-to-source-expression" algorithm>
...
 
-      3.  If the first character of |expression|'s <a grammar>`host-part`</a>
-          is an U+002A ASTERISK character (`*`):
+      3.  If [[#match-hosts]] returns "`Does Not Match`" given |expression|'s 
+          <a grammar>`host-part`</a> and |url|'s {{URL/host}}, return "`Does 
+          Not Match`".
 
-          1.  Let |remaining| be the result of removing the leading "`*`" from
-              |expression|.
+      4.  If [[#match-ports]] returns "`Does Not Match`" given |expression|'s
```
Comment (mikewest) [inline index.src.html]: You need to pass in something like `null` if there's no `port-part`.

File: index.src.html [hunk 3]
```diff
@@ -3493,6 +3433,103 @@ <h5 id="match-schemes" algorithm>
...
+  Note: This algorithm is not transitive. That is, A matching B does not mean 
+  that B will match A.
+
+  1.  If A is empty:
+
+      1.  If B is the <a>default port</a> for Scheme-B, return "`Matches`".
+          
+      2.  Return "`Does Not Match`".
+
+  2.  If A is equal to ":*", return "`Matches`".
```
Comment (mikewest) [inline index.src.html]: Let's change the `host-source` grammar to exclude the `:` from `port-source`. That is, something like:


```
host-source = [ scheme-part "://" ] host-part [ ":" port-part ] [ path-part ]
...
port-part = 1*DIGIT / "*"
```

Otherwise, the matching in 3 is not going to work.

File: index.src.html [hunk 4]
```diff
@@ -3493,6 +3433,103 @@ <h5 id="match-schemes" algorithm>
...
+
+  3.  If A is equal to B, return "`Matches`".
```
Comment (mikewest) [inline index.src.html]: `If |A| is a <a>case-sensitive</a> match for |B|, ...`

File: index.src.html (hunk 1 above)
Comment (mikewest) [inline index.src.html]: The goal here is to ensure that the expression `example.com:80` matches the URL `https://example.com:443/`. We wouldn't want `example.com:80` to match `ftp://example.com:21` or `gopher://example.com:70`.

File: index.src.html [hunk 5]
```diff
@@ -3493,6 +3433,103 @@ <h5 id="match-schemes" algorithm>
...
+      Note: A future version of this specification may allow literal IPv6
+      and IPv4 addresses, depending on usage and demand. Given the weak
+      security properties of IP addresses in relation to named hosts,
+      however, authors are encouraged to prefer the latter whenever
+      possible.
+
+  4.  Return "`Does Not Match`".
+
+  <h5 id="match-ports" algorithm>
+    Does |port-part| match another |port-part|?
```
Comment (mikewest) [inline index.src.html]: "... for a given scheme"

File: index.src.html [hunk 6]
```diff
@@ -3371,79 +3371,19 @@ <h5 id="match-url-to-source-expression" algorithm>
...
-              |expression|.
+      4.  If [[#match-ports]] returns "`Does Not Match`" given |expression|'s 
+          <a grammar>`port-part`</a> and |url|'s {{URL/port}}, return "`Does 
+          Not Match`".
 
-          2.  If |remaining| (including the leading U+002E FULL STOP character
-              (`.`)) is not an <a>ASCII case-insensitive</a> match for the
-              rightmost characters of |url|'s {{URL/host}}, then return "`Does
-              Not Match`".
+      5.  If [[#match-paths]] returns "`Does Not Match`" given |expression|'s
```
Comment (mikewest) [inline index.src.html]: Where did the "redirect count" bit go? We should skip path matching for redirects.

Comment (mikewest) [inline index.src.html]: You need to pass in something like `null` if there's no `path-part`.

File: index.src.html [hunk 7]
```diff
@@ -3371,79 +3371,19 @@ <h5 id="match-url-to-source-expression" algorithm>
...
+      4.  If [[#match-ports]] returns "`Does Not Match`" given |expression|'s 
+          <a grammar>`port-part`</a> and |url|'s {{URL/port}}, return "`Does 
+          Not Match`".
 
-          2.  If |remaining| (including the leading U+002E FULL STOP character
-              (`.`)) is not an <a>ASCII case-insensitive</a> match for the
-              rightmost characters of |url|'s {{URL/host}}, then return "`Does
-              Not Match`".
+      5.  If [[#match-paths]] returns "`Does Not Match`" given |expression|'s 
+          <a grammar>`path-part`</a> and |url|'s {{URL/path}}, return "`Does
```
Comment (mikewest) [inline index.src.html]: `{{URL/path}}` is a list, not a string. If you want to do string comparisons, you need to join the list together with `/`s. That's basically the opposite of the algorithm today, but whatever. :)

**Comment by Sun77789** [comment]
@mikewest updated!

File: index.src.html [hunk 8]
```diff
@@ -3344,106 +3344,38 @@ <h5 id="match-url-to-source-expression" algorithm>
...
-              character (`/`), and `true` otherwise.
-
-          2.  Let |path list| be the result of
-              <a lt="strictly split a string">strictly splitting</a>
-              |expression|'s <a grammar>`path-part`</a> on the U+002F SOLIDUS
-              character (`/`).
-
-          3.  If |path list| has more items than |url|'s {{URL/path}}, return
-              "`Does Not Match`".
+          1.  Let |path| be the resulting <a>ASCII string</a> of joining |url|'s
```
Comment (mikewest) [inline index.src.html]: URLs aren't always ASCII. http://emojipedia.org/💩/, for instance, is totally valid. I'd rephrase this to `Let |path| be the result of joining ...`.

File: index.src.html [hunk 9]
```diff
@@ -3471,28 +3403,135 @@ <h5 id="match-schemes" algorithm>
...
   </h5>
 
-  Given two <a>ASCII strings</a> A and B, this algorithm returns "`Matches`" if A 
-  matches B, and returns "`Does Not Match`" otherwise.
+  Given two <a>ASCII strings</a> (|A| and |B|), this algorithm returns 
+  "`Matches`" if |A| matches |B|, and returns "`Does Not Match`" otherwise.
 
-  Note: This algorithm is not transitive. That is, A matching B does not mean 
-  that B will match A.
+  Note: The matching relation is assymetric. That is, |A| matching |B| does not
```
Comment (mikewest) [inline index.src.html]: Thanks for addressing this. Sorry I told you to use the wrong word. I dumb. :(

File: index.src.html [hunk 10]
```diff
@@ -3471,28 +3403,135 @@ <h5 id="match-schemes" algorithm>
...
+          is an <a>ASCII case-insensitive</a> match for "`https`".
 
   2.  Return "`Does Not Match`".
 
+  Note: This logic effectively means that `script-src http:` is
+  equivalent to `script-src http: https:`, and
+  `script-src http://example.com/` is equivalent to `script-src
+  http://example.com https://example.com`. As well as WebSocket
+  schemes are equivalent to corresponding HTTP schemes. In short,
+  we always allow a secure upgrade from an explicitly insecure expression.
```
Comment (mikewest) [inline index.src.html]: Nit: Let's rephrase this note a bit while we're here:

> Note: We always allow a secure upgrade from an explicitly insecure expression. That is, `script-src http:` is treated as equivalent to `script-src http: https:`, `script-src http://example.com` to `script-src http://example.com https://example.com`, and `connect-src ws:` to `connect-src ws: wss:`.

Could you also add something like:

> ISSUE: Add a reference to https://github.com/diracdeltas/sniffly.

File: index.src.html [hunk 11]
```diff
@@ -3471,28 +3403,135 @@ <h5 id="match-schemes" algorithm>
...
+      possible.
+
+  4.  Return "`Matches`".
+
+  <h5 id="match-ports" algorithm>
+    Does |port-part| match another |port-part| for a given `scheme`?
+  </h5>
+
+  Given three <a>ASCII strings</a> |port A|, |port B|, and |scheme B|, this 
+  algorithm returns "`Matches`" if |A| matches |B|, and returns
```
Comment (mikewest) [inline index.src.html]: Nit: `A` and `B` are variables that you never use again. How about something like '... returns "`Matches`" if |port A| matches |port B| for |scheme B|, and "`Does Not Match`" otherwise.'

File: index.src.html [hunk 12]
```diff
@@ -3471,28 +3403,135 @@ <h5 id="match-schemes" algorithm>
...
+  Note: The matching relation is assymetric. That is, |A| matching |B| does not
+  mean that |B| will match |A|.
+
+  1.  If |A| is empty, return "`Matches`".
+
+  2.  Let |exact match| be `false` if the final character of |A| is the U+002F 
+      SOLIDUS character (`/`), and `true` otherwise.
+
+  3.  Let |path list A| and |path list B| be the result of 
+      <a lt="strictly split |a| string">strictly splitting</a> |A| and |B|
```
Comment (mikewest) [inline index.src.html]: `s/split |a| string/split a string/`

File: index.src.html (hunk 8 above)
Comment (mikewest) [inline index.src.html]: Bah. Nevermind. I checked URL: it percent encodes the path. Sorry!

File: index.src.html [hunk 13]
```diff
@@ -3471,28 +3403,135 @@ <h5 id="match-schemes" algorithm>
...
+      <a lt="strictly split |a| string">strictly splitting</a> |A| and |B| 
+      respestively on the U+002F SOLIDUS character (`/`).
+
+  4.  If |path list A| has more items than |path list B|, return 
+      "`Does Not Match`".
+
+  5.  If |exact match| is `true`, and |path list A| does not have the same
+      number of items as |path list B|, return "`Does Not Match`".
+
+  6.  For each |expression piece| in |path list|:
```
Comment (mikewest) [inline index.src.html]: `s/|expression piece|/|piece A|/`
`s/|path list|/|path list A|/`

File: index.src.html [hunk 14]
```diff
@@ -3471,28 +3403,135 @@ <h5 id="match-schemes" algorithm>
...
+
+      1.  Let |url piece| be the next item in |path list B|.
```
Comment (mikewest) [inline index.src.html]: `s/|url piece|/|piece B|/`

File: index.src.html [hunk 15]
```diff
@@ -3471,28 +3403,135 @@ <h5 id="match-schemes" algorithm>
...
+
+  4.  If |port B| is empty:
+
+      1.  If |port A| is the <a>default port</a> for |scheme B|, return 
+          "`Matches`". Otherwise, return "`Does not Match`".
+
+  5. Return "`Does Not Match`".
+
+  <h5 id="match-paths" algorithm>
+    Does |path-part| match another |path-part|?
```
Comment (mikewest) [inline index.src.html]: How about "Does `path-part` |A| match `path-part` |B|?"

Same thing for the other algorithm names: the `|...|` syntax defines a variable. We should only use it for things we're actually using in the algorithms,

File: index.src.html (hunk 8 above)
Comment (Sun77789) [inline index.src.html]: Great to learn about the URLs and emojipedia :)

File: index.src.html [hunk 16]
```diff
@@ -3344,154 +3344,192 @@ <h5 id="match-url-to-source-expression" algorithm>
...
-              rightmost characters of |url|'s {{URL/host}}, then return "`Does
-              Not Match`".
+      5.  If |expression| contains a non-empty <a grammar>`path-part`</a>, and
+          |redirect count| is 0, then:
 
-      4.  If the first character of |expression|'s <a grammar>`host-part`</a>
-          is not an U+002A ASTERISK character (`*`), and |url|'s {{URL/host}}
-          is not an <a>ASCII case-insensitive</a> match for |expression|'s
-          <a grammar>`host-part`</a>, return "`Does Not Match`".
+          1.  Let |path| be the resulting <a>ASCII string</a> of joining |url|'s
```
Comment (mikewest) [inline index.src.html]: Please change this to "string of <a>ASCII bytes</a>". Sorry; I noticed in https://github.com/w3c/webappsec-csp/commit/9910a0d194c2a45f0d00dfd4e6d5c03cf2fb4bb9 that `<a>ASCII string</a>` is only used in URL, and probably isn't the canonical term we should be using.

Or just avoid the whole question by saying "Let |path| be the result of joining...".

File: index.src.html [hunk 17]
```diff
@@ -3344,154 +3344,192 @@ <h5 id="match-url-to-source-expression" algorithm>
...
-              <a lt="strictly split a string">strictly splitting</a>
-              |expression|'s <a grammar>`path-part`</a> on the U+002F SOLIDUS
-              character (`/`).
+  <h5 id="match-schemes" algorithm>
+    Does `scheme-part` |A| match `scheme-part` |B|?
+  </h5>
 
-          3.  If |path list| has more items than |url|'s {{URL/path}}, return
-              "`Does Not Match`".
+  Given two <a>ASCII strings</a> (|A| and |B|), this algorithm returns
```
Comment (mikewest) [inline index.src.html]: I changed this in https://github.com/w3c/webappsec-csp/commit/9910a0d194c2a45f0d00dfd4e6d5c03cf2fb4bb9. You'll need to rebase/merge this onto ToT.

File: index.src.html [hunk 18]
```diff
@@ -3344,154 +3344,192 @@ <h5 id="match-url-to-source-expression" algorithm>
...
-      1.  |origin| is the same as |url|'s {{URL/origin}}
+  <h5 id="match-hosts" algorithm>
+    Does `host-part` |A| match `host-part` |B|?
+  </h5>
 
-      2.  |origin|'s {{URL/host}} is the same as |url|'s {{URL/host}},
-          |origin|'s {{URL/port}} and |url|'s {{URL/port} are either the same
-          or the <a>default ports</a> for their respective {{URL/scheme}}s, and
-          one or more of the following conditions is met:
+  Given two <a>ASCII strings</a> (|A| and |B|), this algorithm returns
```
Comment (mikewest) [inline index.src.html]: Ditto.

File: index.src.html [hunk 19]
```diff
@@ -3344,154 +3344,192 @@ <h5 id="match-url-to-source-expression" algorithm>
...
+
+  4.  Return "`Matches`".
+
+  <h5 id="match-ports" algorithm>
+    Does `port-part` |A| match `port-part` |B| for |scheme B|?
   </h5>
 
-  Given two <a>ASCII strings</a> A and B, this algorithm returns "`Matches`" if A 
-  matches B, and returns "`Does Not Match`" otherwise.
+  Given three <a>ASCII strings</a> |port A|, |port B|, and |scheme B|, this
```
Comment (mikewest) [inline index.src.html]: Ditto.

File: index.src.html [hunk 20]
```diff
@@ -3344,154 +3344,192 @@ <h5 id="match-url-to-source-expression" algorithm>
...
+      1.  If |port A| is the <a>default port</a> for |scheme B|, return 
+          "`Matches`". Otherwise, return "`Does not Match`".
+
+  5. Return "`Does Not Match`".
+
+  <h5 id="match-paths" algorithm>
+    Does `path-part` |A| match `path-part` |B|?
+  </h5>
+
+  Given two <a>ASCII strings</a> (|A| and |B|), this algorithm returns
```
Comment (mikewest) [inline index.src.html]: Ditto.
